### ISSUE TITLE
PropTypes and Typescript duplication

### ISSUE THREAD

**Issue by elycheea** [carbon-design-system/ibm-products#4466]
From Feb 27, 2024 retrospective.

Currently, we’re keeping both PropTypes and Typescript, but consider this is also duplication.
(Note that Typescript checks type at compile while PropTypes checks during runtime.)
Duplication does also mean additional work for us longer term if we add or change props. 

![Image](https://github.com/carbon-design-system/ibm-products/assets/1717398/cd8d4e37-183a-40f9-9cb2-3dbd4075a9c2)

**Comment by sstrubberg** [comment]
I'd love to have a unified Carbon stance on this topic. We should all get together. @elycheea @jeffchew @tay1orjones @kennylam @matthewgallo @lee-chase 

Maybe a topic for our next Dev call?

**Comment by sstrubberg** [comment]
Slack convo excerpts with Taylor.

- I don't think it's a good idea to remove proptypes right now. There are apps out there not using typescript and therefore relying on PropTypes. This is a play to maximize our compatibility.
- If we were to remove them, we would need to gather data that suggests doing so is the right play.
- It would qualify as a breaking change.

**Comment by elycheea** [comment]
@sstrubberg Sounds like next dev call will have some funteresting topics lined up. 😀 We had a few other we wanted to add to the list.

**Comment by matthewgallo** [comment]
I'm wondering how or if our viewpoint on this changes at all when taking into consideration that `prop-types` is deprecated and will be officially removed in React 19 ([more info here](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#removed-proptypes-and-defaultprops)). Their before and after example from the React 19 upgrade guide shows only typescript types.

```jsx
// Before
import PropTypes from 'prop-types';

function Heading({text}) {
  return <h1>{text}</h1>;
}
Heading.propTypes = {
  text: PropTypes.string,
};
Heading.defaultProps = {
  text: 'Hello, world!',
};
```
-------
```jsx
// After
interface Props {
  text?: string;
}
function Heading({text = 'Hello, world!'}: Props) {
  return <h1>{text}</h1>;
}
```

@tay1orjones I like the idea you posed around possibly generating prop types so that we can maintain support for previous React versions and it would also allow us to remove some of the duplication that we already see in our components.

**Comment by wkeese** [comment]
The other thing is that having PropTypes bloats the code (i.e. it bloats the output from Webpack).   So I understand your desire to be nice to people still using plain Javascript, but it does add a cost to all applications.

Of course this applies to @carbon/react as well as @carbon/ibm-products.

**Comment by tay1orjones** [comment]
@wkeese Production builds should use something like https://www.npmjs.com/package/babel-plugin-transform-react-remove-prop-types to strip the proptypes. 

I'm not sure if the additional size realistically impacts dx during local development.

**Comment by wkeese** [comment]
> @wkeese Production builds should use something like https://www.npmjs.com/package/babel-plugin-transform-react-remove-prop-types to strip the proptypes.

Yes, you should do that in your production build of Carbon.

But it doesn't matter if _I_ run that plugin in _my_ production build, because that would affect the code in Carbon.   In other words, Babel runs on code in the src/ directory, not in the node_modules/ directory.    (I know technically you can configure Babel to run on node_modules/, but it's not practical.)

**Comment by tay1orjones** [comment]
Yeah, we can't strip them but we could wrap them in a `__DEV__` check like we do in `@carbon/icons-react` so they'll be tree shaken/dead code eliminated. It might help us get to a flat entrypoint at some point, too - https://github.com/carbon-design-system/carbon/issues/5442

**Comment by wkeese** [comment]
Sounds good although I'm not actually sure what you mean by "flat entrypoint".

**Comment by wkeese** [comment]
Sounds good although I'm not actually sure what you mean by "flat entrypoint".
